controllers/constraintstatus_controller.go

```diff
@@ -0,0 +1,274 @@
+					obj := e.Object.(*v1beta1.ConstraintPodStatus)
+
+					return slices.Contains(obj.Status.Operations, "audit")
+				},
+			},
+			)).
+		Complete(r)
+}
+
+// User set gatekeeper.spec.audit.auditFromCache to Automatic, this reconcile function
+// collect all constraints and add found kinds which is used in contraint to config.spec.sync.syncOnly
+func (r *ConstraintStatusReconciler) Reconcile(ctx context.Context,
+	request reconcile.Request,
+) (reconcile.Result, error) {
+	log := r.Log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	log.Info("Reconciling ConstraintPodStatus and Config")
+
+	gatekeeper := &operatorv1alpha1.Gatekeeper{}
+	err := r.Get(ctx, types.NamespacedName{
+		Namespace: "",
+		Name:      "gatekeeper",
+	}, gatekeeper)
+
+	auditFromCache := *gatekeeper.Spec.Audit.AuditFromCache
+
+	if err != nil || !strings.EqualFold(string(auditFromCache), "Automatic") {
+		return reconcile.Result{Requeue: false}, fmt.Errorf("auditFromCache is not set so skip this reconcile")
+	}
+
+	// Get config or create if not exist
+	config := &v1alpha1.Config{}
+	err = r.Get(ctx, types.NamespacedName{
+		Namespace: r.Namespace,
+		Name:      "config",
+	}, config)
+
+	if apierrors.IsNotFound(err) {
+		config = &v1alpha1.Config{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "config",
+				Namespace: r.Namespace,
+			},
+			Spec: v1alpha1.ConfigSpec{
+				Validation: v1alpha1.Validation{},
+				Sync: v1alpha1.Sync{
+					SyncOnly: []v1alpha1.SyncOnlyEntry{},
+				},
+			},
+		}
+		err = r.Create(ctx, config)
+
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	contraintStatusList := &v1beta1.ConstraintPodStatusList{}
+	err = r.List(ctx, contraintStatusList, client.InNamespace(request.NamespacedName.Namespace))
+
+	if err != nil {
+		return reconcile.Result{Requeue: false}, err
+	}
+
+	constraintKindNames := map[string][]string{}
+
+	for _, c := range contraintStatusList.Items {
+		if !slices.Contains(c.Status.Operations, "audit") {
+			continue
+		}
+
+		labels := c.GetLabels()
+
+		constraintKindNames[labels["internal.gatekeeper.sh/constraint-kind"]] = append(constraintKindNames[labels["internal.gatekeeper.sh/constraint-kind"]], ////nolint:lll
+			labels["internal.gatekeeper.sh/constraint-name"])
+	}
+
+	syncOnlys := []v1alpha1.SyncOnlyEntry{}
+
+	for kind, names := range constraintKindNames {
+		constraintGVR := schema.GroupVersionResource{
+			Group:    "constraints.gatekeeper.sh",
+			Version:  "v1beta1",
+			Resource: strings.ToLower(kind),
+		}
+		for _, name := range names {
+			constraint, err := r.DynamicClient.Resource(constraintGVR).Get(ctx, name, metav1.GetOptions{})
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+
+			constraintKinds, _, err := unstructured.NestedSlice(constraint.Object, "spec", "match", "kinds")
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+
+			r.addSyncOnlys(constraintKinds, &syncOnlys)
```

Comment:
It'd be nice if you called `addSyncOnlys` only on kinds that haven't been processed already since I imagine there could be many constraints working with the same resources.